src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs

```diff
@@ -680,32 +682,136 @@ public bool Equals([NotNullWhen(true)] string? value, StringComparison compariso
+            static bool EqualsUnrolled_0_to_8(string? a, string b)
             {
-                return a != null && a.Length == 0;
-            }
+                if (a != null)
+                {
+                    if (b.Length == 0)
+                    {
+                        // Fold Equals(a, "") to 'a != null && a.Length == 0'
+                        return a.Length == 0;
+                    }
 
-            if (object.ReferenceEquals(a, b))
-            {
-                return true;
+                    // For Length 1 and 2 we rely on fact that even an empty string is 8 bytes long (on 64bit)
+                    // [ 4b Length ][ 2b _firstChar ][ 2b padding ]
+                    // so we can check Length and first 2 chars in a single operation
+                    if (b.Length == 1)
+                    {
+                        return ReadUInt64(a, -2) == (((ulong)b[0] << 32) | 1UL);
+                    }
+                    if (b.Length == 2)
+                    {
+                        return ReadUInt64(a, -2) == (((ulong)b[1] << 48) |
+                                                     ((ulong)b[0] << 32) | 2UL);
+                    }
+
+                    if (b.Length == 3)
+                    {
+                        // Load ch1, ch2, ch3 and \0 into ulong
+                        return a.Length == 3 &&
+                            ReadUInt64(a) == (((ulong)b[2] << 32) |
+                                              ((ulong)b[1] << 16) | b[0]);
+                    }
+
+                    ulong v1 = ReadUInt64(a);
+                    ulong cns1 = ((ulong)b[3] << 48) |
+                                 ((ulong)b[2] << 32) |
+                                 ((ulong)b[1] << 16) | b[0];
+
+                    if (b.Length == 4)
+                    {
+                        // Load ch1, ch2, ch3 and ch4 into ulong
+                        return a.Length == 4 && v1 == cns1;
+                    }
+                    // Handle Length [5..8] via two ulong (overlapped)
+                    return v1 == cns1 && ReadUInt64(a, a.Length - 4) ==
+                        (((ulong)b[b.Length - 1] << 48) |
+                         ((ulong)b[b.Length - 2] << 32) |
+                         ((ulong)b[b.Length - 3] << 16) |
+                         ((ulong)b[b.Length - 4] << 0));
+                }
+
+                // a is null when b is a known non-null
+                return false;
             }
 
-            if (a is null || b is null || a.Length != b.Length)
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            static bool EqualsUnrolled_9_to_16(string? a, string b)
             {
+                if (a != null && a.Length == b.Length)
+                {
+                    // Load 'a' into two vectors with overlapping.
+                    Vector128<ushort> v2 = Vector128.LoadUnsafe(
+                        ref Unsafe.As<char, ushort>(ref a._firstChar), (nuint)a.Length - 8);
```

Comment:
Is the JIT able to treat `a.Length` as a constant here? Would it be better to use `b.Length`?

Reply:
Good point, thanks!
Unfortunately, JIT is not yet smart enough to handle [this](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQEwEYCwAUEZgMwAEO5AwkQN5HlPkD0L5ApgB4AOHAxgg4ATcvwD2wjgHMOMEOUbM2Spm2b8IAGy3kA6gCcoQgDKwOACgAs2AJQBuVa3ZMDHBE5WFmlXADZKKxptLQ4DC1gEcghbcgBeAD5gnQ5LCDhyGwcnJ0x/QOTQy0jojJLgWycGbx9mKAAzcgsIeLjyCqda5jyATmbsmqYAXyIhoA===A)